src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1433 @@
+                var normalizedPattern = PatternNormalizer.Rewrite(pattern, rootIdentifier.Type);
+                analyze(normalizedPattern, context);
+
+                // The set of redundancies in pattern `A` is identical to the set of redundancies in pattern `not A`
+                // but our method for detecting redundancies only detects those in `or` cases (which we bring to the top after normalization).
+                // By analyzing `not A` too we can report more redundancies.
+                // For example: `if (o is not (<any pattern including a redundancy>))`, `if (i is 42 and not 43)` (which is the negation of `not 42 or 43`)
+                var negated = new BoundNegatedPattern(pattern.Syntax, negated: pattern, pattern.InputType, narrowedType: pattern.InputType);
+                var normalizedNegatedPattern = PatternNormalizer.Rewrite(negated, rootIdentifier.Type);
+                analyze(normalizedNegatedPattern, context);
+            }
+            catch (InsufficientExecutionStackException)
+            {
+                // We silently stop reporting diagnostics for redundant patterns
+            }
+
+            return;
+
+            static void populateStateForCases(ArrayBuilder<BoundPattern> set, PooledHashSet<LabelSymbol> labelsToIgnore,
+                ref TemporaryArray<StateForCase> casesBuilder, ReachabilityAnalysisContext context)
+            {
+                int patternIndex = context.PatternIndex;
+                var previousCases = context.PreviousCases;
+                for (int i = 0; i < patternIndex; i++)
+                {
+                    casesBuilder.Add(previousCases[i]);
+                }
+
+                int index = patternIndex;
+                foreach (BoundPattern pattern in set)
+                {
+                    var label = new GeneratedLabelSymbol("orCase");
+                    SyntaxNode? diagSyntax = pattern.Syntax;
+                    if (pattern.WasCompilerGenerated)
+                    {
+                        labelsToIgnore.Add(label);
+                        diagSyntax = context.Syntax;
+                    }
+
+                    Debug.Assert(diagSyntax is not null);
+                    casesBuilder.Add(context.Builder.MakeTestsForPattern(++index, diagSyntax, context.RootIdentifier, pattern, whenClause: null, label: label));
+                }
+            }
+
+            // We need to reduce the break introduced by reporting redundant patterns
+            // and we never want to affect people who express their patterns thoroughly (but correctly)
+            // such as `switch { < 0 => -1, 0 => 0, > 0 => 1 }`
+            // So we're only reporting a warning for situations that syntactically look hazardous.
+            // Others are reported as a hidden diagnostic.
+            // At the moment, we're only interested in patterns in an `or` pattern with a `not` in an early case of the `or`.
+            static bool detectNotOrPattern(SyntaxNode syntax)
```

Comment:
>detectNotOrPattern [](http://example.com/codeflow?start=24&length=18)

Without clearly stated description of the algorithm that we are trying to implement in this method, I find it very hard to confirm that the implementation is complete. I could be wrong, but it feels like the implementation is lacking some generality. It feels ad hoc, it feels like the implementation is adjusted to suite specific scenarios that were tried out, rather than following a general plan. This makes me worry that the implementation either might be incomplete (won't behave properly for scenarios we haven't tried), or unnecessary complicated/redundant in some aspects. Let me state the algorithm which I think we are trying to (intend to) implement:
We start with a node of interest
1. If the node of interest is on the right side of `or` and its left side contains a negated pattern in the "right" position, return `true`. What does it mean to contain a negated pattern in the "right" position needs to be clarified separately.  
2. If the node of interest is an immediate child of `ParenthesizedPatternSyntax`, make the `ParenthesizedPatternSyntax` the node of interest and restart from step 1.
3. If the node of interest is an immediate child of `BinaryPatternSyntax`, make the `BinaryPatternSyntax` the node of interest and restart from step 1. 
4. If the node of interest is a pattern in a sub-pattern of a recursive pattern, make the recursive pattern the node of interest and restart from step 1. 
5. If the node of interest is a pattern in a list pattern, make the list pattern the node of interest and restart from step 1. 
6. If the node of interest is a pattern in a slice pattern, make the list pattern the node of interest and restart from step 1. 
7. Deal with pattern in ITuple pattern similarly, if we decide that is the right thing to do
8. Return false.
 
What does it mean to contain a negated pattern in the "right" position within the left child of an `or ?
We start with a node of interest
1. If the node of interest is a NotPattern, return true.
2. If the node of interest is an `or` pattern, return true if the process returns true for either child.
3. If the node of interest is an `and` pattern, return true if any of its children contain a negated pattern in the "right" position
4. Return false

What does it mean to contain a negated pattern in the "right" position for a child of an `and ?
We start with a node of interest
1. If the node of interest is a NotPattern, return true.
2. If the node of interest is an `and` pattern, return true if the process returns true for either child.
3. Return false

If we agree that the described algorithm accurately reflects the logic that we intent to implement, then
simply implementing the provided steps will result in a simpler and easier to follow code. The code that
is guaranteed to be correct.